lib/std/math/mean.zig

```diff
@@ -0,0 +1,132 @@
+//! Arithmetic mean (https://mathworld.wolfram.com/ArithmeticMean.html)
+const std = @import("std");
+const math = std.math;
+const testing = std.testing;
+
+const max_int_bits = 65535;
+
+/// Computes the mean (average) of a slice of items of type `T`.
+/// The type `T` must be either a floating point type or an integer type.
+/// Avoids overflow by using a wider intermediate type for the sum.
+///
+/// Floating point type f128 could still overflow to infinity. You need to handle that case separately.
+pub fn mean(comptime T: type, items: []const T) T {
+    std.debug.assert(items.len > 0);
+    const info = @typeInfo(T);
+    return switch (info) {
+        .float => float(T, items),
+        .int => int(T, items),
+        else => @compileError("T must be float or integer, found " ++ @typeName(T)),
+    };
+}
+
+fn float(comptime T: type, items: []const T) T {
+    const info = @typeInfo(T);
+    const Wider: type = switch (info.float.bits) {
+        16 => f32,
+        32 => f64,
+        64 => f128,
+        80 => f128,
+        128 => f128,
+        else => @compileError("unknown floating point type " ++ @typeName(T)),
+    };
+    var sum: Wider = 0;
+    for (items) |item| {
+        sum += @as(Wider, item);
+    }
+    const len_wide: Wider = @floatFromInt(items.len);
+    const mean_wide: Wider = sum / len_wide;
+    return @floatCast(mean_wide);
+}
+
+fn int(comptime T: type, items: []const T) T {
+    const info = @typeInfo(T);
+    const sign = info.int.signedness;
+    const bits = info.int.bits;
+    const usize_bits = @bitSizeOf(usize);
+    const Wider: type = switch (max_int_bits - usize_bits >= bits) {
+        true => @Type(.{ .int = .{ .signedness = sign, .bits = bits + usize_bits } }),
+        false => @compileError("Calculation can overflow for type " ++ @typeName(T)),
```

Comment:
Rather than this check for an edge case which ties this code more to current language rules, I'd just let this case compile error as it normally would. Nobody is going to hit it in practice, as nobody really uses `u65535` (that's kinda the point of the integer bit limit being a `u16`, it allows values as high as anyone could realistically want!). If someone somehow does hit it, then the compile error they do get will be perfectly reasonable. So:
```suggestion
    const Wider = @Type(.{ .int = .{ .signedness = sign, .bits = bits + usize_bits } });
```
Oh yeah, also, small style thing: I suggest not including type annotations for the type `type` unless it's unclear for some reason. So, this is `const Wider = ...` instead of `const Wider: type = ...`.